File: network/src/main/java/com/zjf/network/ZRest.java
package com.zjf.network;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.zjf.network.models.INetExternalParams;
import io.reactivex.schedulers.Schedulers;
import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import java.io.File;
import java.io.IOException;
import java.net.Proxy;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * @Desc 公用网络类 初始化okhttpclient和Retrofit
 * @Author zjf
 * @Date 2019/3/15
 */
public class ZRest {

    private final Retrofit mRetrofit;
    private final OkHttpClient mOkHttpClient;
    private static ZRest mpmRest;

    public static ZRest getInstance() {
        if (mpmRest == null) {
            mpmRest = new ZRest();
        }
        return mpmRest;
    }

    private ZRest() {
        final OkHttpClient.Builder builder;
        builder = generateDefaultOkHttpBuilder();
        final INetExternalParams networkParams = ZNetWork.get().externalParams();

        List<Interceptor> interceptorList = ZNetWork.get().interceptors();
        for (Interceptor interceptor : interceptorList) {
            builder.addInterceptor(interceptor);
        }

        List<Interceptor> networkInterceptorsList = ZNetWork.get().networkInterceptors();
        for (Interceptor interceptor : networkInterceptorsList) {
            builder.addNetworkInterceptor(interceptor);
        }
        builder.proxy(getProxy(networkParams));
        mOkHttpClient = builder.build();

        Gson gson = new GsonBuilder()
                .setLenient()
                .registerTypeAdapterFactory(new ItemTypeAdapterFactory())
                .create();

        mRetrofit = new Retrofit.Builder()
                .baseUrl(networkParams.baseUrl())
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxTransformErrorCallAdapterFactory.createWithScheduler(Schedulers.io()))
                .client(mOkHttpClient)
                .validateEagerly(true)
                .build();
    }



    private Proxy getProxy(INetExternalParams iNetExternalParams) {
        if (!iNetExternalParams.isRelease()) {
            //非release，由proxySelector来选择
            return null;
        }
        return Proxy.NO_PROXY;
    }

    public OkHttpClient getOkHttpClient() {
        return mOkHttpClient;
    }

    public Retrofit getRetrofit() {
        return mRetrofit;
    }

//    public OkHttpClient getDownLoadHttpClient(OnDownloadListener onDownloadListener) {
//        OkHttpClient.Builder  builder = generateDefaultOkHttpBuilder();
//        //创建一个OkHttpClient，并添加网络拦截器
//        OkHttpClient client = builder
//                .addNetworkInterceptor(new Interceptor() {
//                    @Override
//                    public Response intercept(Chain chain) throws IOException {
//                        Response response = chain.proceed(chain.request());
//                        //这里将ResponseBody包装成我们的ProgressResponseBody
//                        return response.newBuilder()
//                                .body(new ProgressResponseBody(response.body(),onDownloadListener))
//                                .build();
//                    }
//                })
//                .build();
//        return client;
//    }

    @SuppressWarnings("all")
    public <T> T create(Class<T> service) {
        return mRetrofit.create(service);
    }

    private static final int CONNECT_TIMEOUT = 50;
    private static final int READ_TIMEOUT = 60;
    private static final int WRITE_TIMEOUT = 60;

    private static Cache getCache() {
        //缓存文件夹
        File cacheFile = new File(ZNetWork.get().mApplicationContext.getExternalCacheDir().toString(), "mpm/reponse");
        //缓存大小为200M
        int cacheSize = 100 * 1024 * 1024;
        //创建缓存对象
        return new Cache(cacheFile, cacheSize);
    }

    private static OkHttpClient.Builder generateDefaultOkHttpBuilder() {
        return new OkHttpClient.Builder()
//                .cache(getCache())
                .writeTimeout(WRITE_TIMEOUT, TimeUnit.SECONDS)
                .connectTimeout(CONNECT_TIMEOUT, TimeUnit.SECONDS)
                .readTimeout(READ_TIMEOUT, TimeUnit.SECONDS);
    }
}

File: network/src/main/java/com/zjf/network/ZRest.kt
package com.zjf.network

import com.google.gson.GsonBuilder
import io.reactivex.schedulers.Schedulers
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

/**
 * @Desc 初始化retrofit和okhttp
 * @Author zjf
 * @Date 2019/3/22
 */
class ZRest {

    private var okHttpClient: OkHttpClient? = null
    private var mRetrofit: Retrofit? = null

    companion object {
        private var zRest: ZRest? = null
        private const val CONNECT_TIMEOUT = 15
        private const val READ_TIMEOUT = 30
        private const val WRITE_TIMEOUT = 30

        fun getInstance() : ZRest? {
            if (zRest == null) {
                zRest = ZRest()
            }
            return zRest
        }
    }

    private constructor() {
        val networkParams = ZNetwork.get()?.externalParams()
        okHttpClient = with(generateDefaultOkHttpBuilder()) {
            ZNetwork.get()?.networkInterceptors()?.let {
                it.forEach {
                    addNetworkInterceptor(it)
                }
            }
            ZNetwork.get()?.interceptors()?.let {
                it.forEach {
                    addInterceptor(it)
                }
            }
            build()
        }
        val gson = GsonBuilder()
            .setLenient()
            .registerTypeAdapterFactory(ItemTypeAdapterFactory())
            .create()

        mRetrofit =  with(Retrofit.Builder())  {
            baseUrl(networkParams?.baseUrl())
            addConverterFactory(GsonConverterFactory.create(gson))
            addCallAdapterFactory(RxTransformErrorCallAdapterFactory.createWithScheduler(Schedulers.io()))
            client(okHttpClient)
            validateEagerly(true)
            build()
        }
    }

    fun getOkHttpClient(): OkHttpClient? {
        return okHttpClient
    }

    fun getRetrofit(): Retrofit? {
        return mRetrofit
    }

    fun <T> create(service: Class<T>): T? {
        return mRetrofit?.create(service)
    }

    private fun generateDefaultOkHttpBuilder(): OkHttpClient.Builder {
        return OkHttpClient.Builder()
            //                .cache(getCache())
            .writeTimeout(WRITE_TIMEOUT.toLong(), TimeUnit.SECONDS)
            .connectTimeout(CONNECT_TIMEOUT.toLong(), TimeUnit.SECONDS)
            .readTimeout(READ_TIMEOUT.toLong(), TimeUnit.SECONDS)
    }
}